### PR TITLE
Adds ability to use default hasher in `dev_mode` for explicit key binding

### LINKED ISSUE #14118: Explicit Key Binding should support default hasher in `dev_mode`
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Motivation

How is the dev_mode syntax supposed to work with the explicit key binding in eg StorageMap?

I mean something like:

```rust
#[pallet::storage]
pub type Baz<T: Config> = StorageMap<Key = T::AccountId, Value = T::Balance>;
```

### Request

Recently, we added the ability to specify hashers as `_` in `dev_mode`, which are replaced by `Blake2_128Concat`. We should also be able to use a default hasher in `dev_mode` in case of explicit key binding.

### Solution

_No response_

### Are you willing to help with this request?

Yes!

### PR DESCRIPTION
**Pull request by gupnik**
Fixes #14118 

Recently, we added the ability to specify hashers as _ in dev_mode, which are replaced by Blake2_128Concat. This PR allows using a default hasher in dev_mode in case of explicit key binding.

With these changes, a syntax like the below works in `dev_mode` but throws an error otherwise.

```rust
#[pallet::storage]
pub type Baz<T: Config> = StorageMap<Key = T::AccountId, Value = T::Balance>;
```

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2848260

File: frame/support/procedural/src/pallet/parse/storage.rs [hunk 1]
```diff
@@ -432,15 +463,15 @@ fn process_named_generics(
 				hasher1: parsed
 					.remove("Hasher1")
 					.map(|binding| binding.ty)
-					.expect("checked above as mandatory generic"),
+					.unwrap_or(syn::Type::Verbatim(quote::quote! { Blake2_128Concat })),
```
Comment (KiChjang) [inline frame/support/procedural/src/pallet/parse/storage.rs:466]: I think you can do this:

```suggestion
					.unwrap_or(syn::parse_quote! { Blake2_128Concat }),
```

File: frame/support/procedural/src/pallet/parse/storage.rs [hunk 2]
```diff
@@ -432,15 +463,15 @@ fn process_named_generics(
...
 				key1: parsed
 					.remove("Key1")
 					.map(|binding| binding.ty)
 					.expect("checked above as mandatory generic"),
 				hasher2: parsed
 					.remove("Hasher2")
 					.map(|binding| binding.ty)
-					.expect("checked above as mandatory generic"),
+					.unwrap_or(syn::Type::Verbatim(quote::quote! { Blake2_128Concat })),
```
Comment (sam0x17) [inline frame/support/procedural/src/pallet/parse/storage.rs:474]: same here:

```suggestion
					.unwrap_or(syn::parse_quote!(Blake2_128Concat)),
```

**Review by sam0x17 — APPROVED** [review]
looks good to me once those two are changed to `parse_quote!`

File: frame/support/procedural/src/pallet/parse/storage.rs [hunk 3]
```diff
@@ -393,10 +404,20 @@ fn process_named_generics(
...
 		StorageKind::CountedMap => {
+			let mandatory_generics: &[&str];
+			let optional_generics: &[&str];
+			if dev_mode {
+				mandatory_generics = &["Key", "Value"];
+				optional_generics = &["Hasher", "QueryKind", "OnEmpty", "MaxValues"];
+			} else {
+				mandatory_generics = &["Hasher", "Key", "Value"];
+				optional_generics = &["QueryKind", "OnEmpty", "MaxValues"];
+			}
```
Comment (ggwpez) [inline frame/support/procedural/src/pallet/parse/storage.rs:415]:
```suggestion
			let mut mandatory_generics = vec!["Key", "Value"];
			let mut optional_generics = vec!["QueryKind", "OnEmpty", "MaxValues"];

			if dev_mode {
				optional_generics.push("Hasher");
			} else {
				mandatory_generics.push("Hasher");
			}
```
Less repetition -> less typos :sparkles:  
(and then use them with a `&` below)

**Review by ggwpez — APPROVED** [review]
Looks good overall :+1:

File: frame/support/test/tests/pallet_ui/dev_mode_explicit_key_without_arg_default_hasher.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#![cfg_attr(not(feature = "std"), no_std)]
```
[the remark this answers is not in the document]
Comment (gupnik) [reply frame/support/test/tests/pallet_ui/dev_mode_explicit_key_without_arg_default_hasher.rs:1]: It actually confirms that the Hasher is needed when not in dev_mode.

[the remark this answers is not in the document]
Comment (sam0x17) [reply frame/support/test/tests/pallet_ui/dev_mode_explicit_key_without_arg_default_hasher.rs:1]: maybe "non_dev_mode.."

**Comment by gupnik** [comment]
bot fmt

**Review by ggwpez — APPROVED** [review]

**Review by KiChjang — APPROVED** [review]

**Comment by gupnik** [comment]
bot merge

**Pull request merged**